core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -237,23 +243,24 @@ public InternalEngine(EngineConfig engineConfig) {
 
     private LocalCheckpointTracker createLocalCheckpointTracker(
         BiFunction<Long, Long, LocalCheckpointTracker> localCheckpointTrackerSupplier) throws IOException {
-        final long maxSeqNo;
-        final long localCheckpoint;
         switch (openMode) {
             case CREATE_INDEX_AND_TRANSLOG:
-                maxSeqNo = SequenceNumbers.NO_OPS_PERFORMED;
-                localCheckpoint = SequenceNumbers.NO_OPS_PERFORMED;
-                break;
-            case OPEN_INDEX_AND_TRANSLOG:
+                return localCheckpointTrackerSupplier.apply(SequenceNumbers.NO_OPS_PERFORMED, SequenceNumbers.NO_OPS_PERFORMED);
             case OPEN_INDEX_CREATE_TRANSLOG:
-                final Tuple<Long, Long> seqNoStats = store.loadSeqNoInfo();
-                maxSeqNo = seqNoStats.v1();
-                localCheckpoint = seqNoStats.v2();
+                final Tuple<Long, Long> seqNoInfo = store.loadSeqNoInfo(null);
+                logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", seqNoInfo.v1(), seqNoInfo.v2());
+                return localCheckpointTrackerSupplier.apply(seqNoInfo.v1(), seqNoInfo.v2());
+            case OPEN_INDEX_AND_TRANSLOG:
+                // When recovering from a previous commit point, we use the local checkpoint from that commit,
+                // but the max_seqno from the last commit. This allows use to throw away stale operations.
+                assert startingCommit != null;
+                final long localCheckpoint = store.loadSeqNoInfo(startingCommit).v2();
+                final long maxSeqNo = store.loadSeqNoInfo(null).v1();
                 logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", maxSeqNo, localCheckpoint);
-                break;
-            default: throw new IllegalArgumentException("unknown type: " + openMode);
+                return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
+            default:
+                throw new IllegalArgumentException("unknown type: " + openMode);
         }
-        return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
```

Comment:
i like the final approach better this gives a uniform return value (easier to understand) and it makes sure these things are set.